// rc-scroll-anim/lib/ScrollOverPack.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _EventDispatcher = require('./EventDispatcher');

var _EventDispatcher2 = _interopRequireDefault(_EventDispatcher);

var _ScrollElement2 = require('./ScrollElement');

var _ScrollElement3 = _interopRequireDefault(_ScrollElement2);

var _util = require('./util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function noop() {}

var ScrollOverPack = function (_ScrollElement) {
  (0, _inherits3['default'])(ScrollOverPack, _ScrollElement);

  function ScrollOverPack(props) {
    (0, _classCallCheck3['default'])(this, ScrollOverPack);

    var _this = (0, _possibleConstructorReturn3['default'])(this, (ScrollOverPack.__proto__ || Object.getPrototypeOf(ScrollOverPack)).call(this, props));

    _this.scrollEventListener = function (e) {
      _this.getParam(e);
      var isTop = _this.elementShowHeight > _this.clientHeight + _this.leavePlayHeight;
      if (_this.enter || !_this.props.replay && isTop) {
        if (!_this.state.show) {
          _this.setState({
            show: true
          });
        }
        if (!_this.props.always && _this.eventType) {
          _EventDispatcher2['default'].removeEventListener(_this.eventType, _this.scrollEventListener, _this.target);
        }
      } else {
        var bottomLeave = _this.elementShowHeight < _this.playHeight;
        // 设置往上时的出场点...
        var topLeave = _this.props.replay ? isTop : null;
        if (topLeave || bottomLeave) {
          if (_this.state.show) {
            _this.setState({
              show: false
            });
          }
        }
      }
    };

    _this.children = (0, _util.toArrayChildren)(props.children);
    _this.oneEnter = false;
    _this.enter = false;
    _this.state = {
      show: false,
      children: (0, _util.toArrayChildren)(props.children)
    };
    return _this;
  }

  (0, _createClass3['default'])(ScrollOverPack, [{
    key: 'render',
    value: function render() {
      var placeholderProps = (0, _objectWithoutProperties3['default'])(this.props, []);

      ['playScale', 'replay', 'component', 'always', 'scrollEvent', 'appear', 'location', 'targetId'].forEach(function (key) {
        return delete placeholderProps[key];
      });
      var childToRender = void 0;
      if (!this.oneEnter) {
        var show = !this.props.appear;
        var children = (0, _util.toArrayChildren)(this.props.children).map(function (item) {
          return item.type.isTweenOne ? _react2['default'].cloneElement(item, (0, _extends3['default'])({}, item.props, { paused: !show })) : _react2['default'].cloneElement(item, item.props, show && item.props.children);
        });
        childToRender = (0, _react.createElement)(this.props.component, (0, _extends3['default'])({}, placeholderProps), children);
        this.oneEnter = true;
      } else {
        if (!this.state.show) {
          this.children = this.children.map(function (item) {
            if (!item) {
              return null;
            }
            // 判断 TweenOne;
            if (item.type.isTweenOne) {
              return _react2['default'].cloneElement(item, { reverse: true });
            }
            return _react2['default'].cloneElement(item, {}, null);
          });
        } else {
          this.children = this.state.children;
        }
        childToRender = (0, _react.createElement)(this.props.component, (0, _extends3['default'])({}, placeholderProps), this.children);
      }
      return childToRender;
    }
  }]);
  return ScrollOverPack;
}(_ScrollElement3['default']);

ScrollOverPack.propTypes = {
  component: _propTypes2['default'].oneOfType([_propTypes2['default'].func, _propTypes2['default'].string]),
  playScale: _propTypes2['default'].any,
  always: _propTypes2['default'].bool,
  scrollEvent: _propTypes2['default'].func,
  children: _propTypes2['default'].any,
  className: _propTypes2['default'].string,
  style: _propTypes2['default'].any,
  replay: _propTypes2['default'].bool,
  onChange: _propTypes2['default'].func,
  appear: _propTypes2['default'].bool
};

ScrollOverPack.defaultProps = {
  component: 'div',
  playScale: 0.5,
  always: true,
  scrollEvent: noop,
  replay: false,
  onChange: noop,
  appear: true
};
ScrollOverPack.isScrollOverPack = true;

exports['default'] = ScrollOverPack;
module.exports = exports['default'];
