// rc-scroll-anim/es/ScrollElement.js
import _extends from 'babel-runtime/helpers/extends';
import _objectWithoutProperties from 'babel-runtime/helpers/objectWithoutProperties';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _createClass from 'babel-runtime/helpers/createClass';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import mapped from './Mapped';
import EventListener from './EventDispatcher';
import { currentScrollTop, transformArguments, windowHeight, toArrayChildren } from './util';

var noop = function noop() {};

var ScrollElement = function (_React$Component) {
  _inherits(ScrollElement, _React$Component);

  function ScrollElement() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, ScrollElement);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = ScrollElement.__proto__ || Object.getPrototypeOf(ScrollElement)).call.apply(_ref, [this].concat(args))), _this), _this.getParam = function (e) {
      _this.clientHeight = _this.target ? _this.target.clientHeight : windowHeight();
      var scrollTop = _this.target ? _this.target.scrollTop : currentScrollTop();
      var domRect = _this.dom.getBoundingClientRect();
      var targetTop = _this.target ? _this.target.getBoundingClientRect().top : 0;
      var offsetTop = domRect.top + scrollTop - targetTop;
      _this.elementShowHeight = scrollTop - offsetTop + _this.clientHeight;
      var playScale = transformArguments(_this.props.playScale);
      var playScaleEnterArray = /([\+\-]?[0-9#\.]+)(px|vh|%)?/.exec(String(playScale[0]));
      if (!playScaleEnterArray[2]) {
        _this.playHeight = _this.clientHeight * parseFloat(playScale[0]);
      } else if (playScaleEnterArray[2] === 'px') {
        _this.playHeight = parseFloat(playScaleEnterArray[1]);
      } else {
        _this.playHeight = _this.clientHeight * parseFloat(playScaleEnterArray[1]) / 100;
      }
      var leaveHeight = domRect.height;
      var playScaleLeaveArray = /([\+\-]?[0-9#\.]+)(px|vh|%)?/.exec(String(playScale[1]));
      if (!playScaleLeaveArray[2]) {
        _this.leavePlayHeight = leaveHeight * parseFloat(playScale[1]);
      } else if (playScaleLeaveArray[2] === 'px') {
        _this.leavePlayHeight = parseFloat(playScaleLeaveArray[1]);
      } else {
        _this.leavePlayHeight = leaveHeight * parseFloat(playScaleLeaveArray[1]) / 100;
      }
      var enter = _this.elementShowHeight >= _this.playHeight && _this.elementShowHeight <= _this.clientHeight + _this.leavePlayHeight;
      var enterOrLeave = enter ? 'enter' : 'leave';
      var mode = _this.enter !== enter || typeof _this.enter !== 'boolean' ? enterOrLeave : null;
      if (mode) {
        _this.props.onChange({ mode: mode, id: _this.props.id }, e);
      }
      _this.enter = enter;
    }, _this.scrollEventListener = function (e) {
      _this.getParam(e);
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(ScrollElement, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.dom = ReactDOM.findDOMNode(this);
      if (this.props.location) {
        this.dom = document.getElementById(this.props.location);
        mapped.register(this.props.location, this.dom);
      } else if (this.props.id) {
        mapped.register(this.props.id, this.dom);
      }
      var date = Date.now();
      this.target = this.props.targetId && document.getElementById(this.props.targetId);

      var length = EventListener._listeners.scroll ? EventListener._listeners.scroll.length : 0;
      this.eventType = 'scroll.scrollEvent' + date + length;
      EventListener.addEventListener(this.eventType, this.scrollEventListener, this.target);

      var scrollTop = currentScrollTop();
      if (!scrollTop) {
        this.scrollEventListener();
      }
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      this.setState({
        children: toArrayChildren(nextProps.children)
      });
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      mapped.unRegister(this.props.id);
      EventListener.removeEventListener(this.eventType, this.scrollEventListener, this.target);
    }
  }, {
    key: 'render',
    value: function render() {
      var props = _objectWithoutProperties(this.props, []);

      ['component', 'playScale', 'location', 'targetId'].forEach(function (key) {
        return delete props[key];
      });
      return React.createElement(this.props.component, _extends({}, props));
    }
  }]);

  return ScrollElement;
}(React.Component);

ScrollElement.propTypes = {
  component: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
  playScale: PropTypes.any,
  id: PropTypes.string,
  onChange: PropTypes.func,
  location: PropTypes.string,
  targetId: PropTypes.string
};

ScrollElement.defaultProps = {
  component: 'div',
  onChange: noop,
  playScale: 0.5
};
ScrollElement.isScrollElement = true;
export default ScrollElement;
